torch/optim/adamax.py

```diff
@@ -287,22 +294,38 @@ def _single_tensor_adamax(
         # Update biased first moment estimate.
         exp_avg.lerp_(grad, 1 - beta1)
         # Update the exponentially weighted infinity norm.
-        if not differentiable:
+        if capturable or differentiable:
+            step = step_t
+
+            bias_correction = 1 - beta1 ** step
+            clr = lr / bias_correction
+            clr_neg = clr.neg()
+
+            if differentiable:
+                norm_buf = torch.cat(
+                    [exp_inf.mul_(beta2).unsqueeze(0), grad.abs().add_(eps).unsqueeze_(0)], 0
+                )
+                exp_inf.copy_(torch.amax(norm_buf, 0, keepdim=False))
+            else:
+                torch.maximum(
+                    exp_inf.mul_(beta2),
+                    grad.abs().add_(eps),
+                    out=exp_inf,
+                )
+            exp_avg = exp_avg.mul(clr_neg)
```

Comment:
@mlazos I noticed here when I used in-place operation the tests failed (assert error different output from the eager optimizer).
Could you please tell me why?

Reply:
You're updating the optimizer state `exp_avg` when you use an inplace operation. At line 328 below, this doesn't occur, so that's why there's a discrepancy. You have to be careful about accidentally updating the optimizer state. Unless updating the state is needed for correctness, we usually only use inplace operations on intermediates (ie tensors that are only local copies to this function)